### PR TITLE
fixes #4243 #4576 #4599 Correctly identify the nodejs environment.

### LINKED ISSUE #4243: "WINDOW NOT DEFINED"... again :-(
**Describe the bug**
Importing the latest aws-amplify into a nuxt application crashes with a "Window not defined" error

**To Reproduce**
In a Nuxt Js application (and probably any SSR framework), try to import the latest aws-amplify

**Expected behavior**
Should not crash

**Additional context**
Most certainly a regression for #3477 

My latest working config is with aws-amplify version 1.1.36
with the following resolutions in package.json:
```json
  "resolutions": {
    "@aws-amplify/analytics": "1.2.23",
    "@aws-amplify/api": "1.0.42",
    "@aws-amplify/auth": "1.2.31",
    "@aws-amplify/cache": "1.0.32",
    "@aws-amplify/core": "1.1.0",
    "@aws-amplify/interactions": "1.0.35",
    "@aws-amplify/predictions": "1.0.3",
    "@aws-amplify/pubsub": "1.1.0",
    "@aws-amplify/storage": "1.1.0"
 }
```

### PR DESCRIPTION
**Pull request by Amplifiyer**
_Issue #, if available:_#4243 #4576 #4599

_Description of changes:_ 

fixes #4243 #4576 #4599 Correctly identify the nodejs environment.

Webpack adds node shim to the bundle overriding the `process` variable
 causing issues detecting nodejs environment. Creating a new function will
 help to avoid incorrect environment detection as new function will have
 it's `this` binded to global scope. Credit: https://stackoverflow.com/a/31090240

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/core/src/JS.ts [hunk 1]
```diff
@@ -165,17 +165,26 @@ export default class JS {
...
-		const isNode =
-			typeof process !== 'undefined' &&
-			process.versions != null &&
-			process.versions.node != null;
+		// function to check if the global scope is "window"
+		const isBrowser = new Function(
+			'try {return this===window;}catch(e){ return false;}'
+		);
+
+		// function to test if global sope is binded to "global"
```
Comment (haverchuck) [inline packages/core/src/JS.ts]: should this be global 'scope'?

**Comment by haverchuck** [comment]
@Amplifiyer How will these checks on this === window or this ==== global be affected if they are using some sort of polyfill for either?

**Comment by Amplifiyer** [comment]
> @Amplifiyer How will these checks on this === window or this ==== global be affected if they are using some sort of polyfill for either?

@haverchuck That's where new Function() helps since it binds to the global scope while polyfills etc would only impact the inner scope. This PR is for this exact reason since webpack was creating polyfills in the bundles and nodejs environments were not able to detect that it's running in Nodejs and causing errors "Not supported"

File: packages/core/src/JS.ts (hunk 1 above)
Comment (Amplifiyer) [reply packages/core/src/JS.ts]: yes, it's a typo, will update it.

**Comment by Amplifiyer** [comment]
> This is pretty clever. Differentiating the environment at runtime within webpack bundles has been a pain-point for a while since `webpack.DefinePlugin` started polyfilling `process.env.NODE_ENV`.
> 
> I don't see anything that sticks out to warrant blocking, but have you considered how this could be tested?
> 
> Jest has support for making the test environment run as `jsdom` (default) or `node`, so it may be possible to have two different test files running within each environment to validate & prevent future regressions:
> 
> > https://jestjs.io/docs/en/configuration#testenvironment-string

@ericclemmons , that's a very good point. I'm struggling to figure out how to test the webpack built bundle (which is where the polyfills are causing issues) with jest. Currently Jest tests the actual code and not the built code. Maybe import from dist if it's the recommended way?

**Comment by ericclemmons** [comment]
@Amplifiyer I know React has success testing against the built code (vs. the source), specifically for this reason.

With the combination of using the `dist` output and setting `@jest-environment jsdom` and `@jest-environment node`, this could be our best option.

Do you think that'd correctly match the scenarios in the issues?

**Comment by Amplifiyer** [comment]
> @Amplifiyer I know React has success testing against the built code (vs. the source), specifically for this reason.
> 
> With the combination of using the `dist` output and setting `@jest-environment jsdom` and `@jest-environment node`, this could be our best option.
> 
> Do you think that'd correctly match the scenarios in the issues?

Yes exactly the scenarios we want to test. I'll look into react tests.

**Comment by Amplifiyer** [comment]
Added unit tests and verified that the node test fails with the current master and passes with this PR.

**Pull request merged**